working/1426-extension-types/feature-specification.md

```diff
@@ -0,0 +1,752 @@
+
+If we introduce any non-object entities in Dart (that is, entities that
+cannot be assigned to a variable of type `Object?`, e.g., external C /
+JavaScript / ... entities, or non-boxed tuples, etc), then we may wish to
+allow for extension types whose on-type is a non-object type.
+
+This should not cause any particular problems: If the on-type is a
+non-object type then the extension type will not be a subtype of `Object`.
+
+
+## Discussion
+
+It would be possible to reify extension types when they occur as type
+arguments of a generic type.
+
+This might help ensuring that the associated discipline of the extension
+type is applied to the elements in, say, a list, even in the case where
+that list is obtained under the type `dynamic`, and a type test or type
+cast is used to confirm that it is a `List<U>` where `U` is an extension
+type.
+
+However, this presumably implies that the cast to a plain `List<T>` where
+`T` is the on-type corresponding to `U` should fail; otherwise the
+protection against accessing the elements using the underlying on-type will
+easily be violated. Moreover, even if we do make this cast fail then we
+could cast each element in the list to `T`, thus still accessing the
+elements using the on-type rather than the more disciplined extension type
+`U`.
+
+We cannot avoid the latter if there is no run-time representation of the
+extension type in the elements in the list, and that is assumed here: For
+example, if we have an instance of `int`, and it is accessed as `extension
+MyInt on int`, the dynamic representation will be a plain `int`, and not
+some wrapped entity that contains information that this particular `int` is
+viewed as a `MyInt`. It seems somewhat inconsistent if we maintain that a
+`List<MyInt>` cannot be viewed as a `List<int>`, but a `MyInt` can be
+viewed as an `int`.
+
+As for promotion, we could consider "promoting to a supertype" when that
+type is an extension type: Assume that `U` is an extension type with
+on-type `T`, and the type of a promotable local variable `x` is `T` or a
+subtype thereof; `x is U` could then _demote_ `x` to have type `U`, even
+though `is` tests normally do not demote. The rationale would be that the
+treatment of `x` as a `U` is conceptually more "informative" and "strict"
+than the treatment of `x` as a `T`, which makes it somewhat similar to a
+downcast.
+
+Note that we can use extension types to handle `void`:
+
+```dart
+extension void on Object? {}
+```
+
+This means that `void` is an "epsilon-supertype" of all top types (it's a
+proper supertype, but just a little bit). It is also a subtype of
+`Object?`, of course, so that creates an equivalence class of "equal" types
+spelled differently. That's a well-known concept today, so we can handle
+that (and it corresponds to the dynamic semantics).
+
+This approach does not admit any member accesses for a receiver of type
+`void`, and it isn't assignable to anything else without a cast. Just like
+`void` of today.
```

Comment:
Now that we mention `void` ...
If we define `extension V on void { int foo() => 42; }`, what can I do with it?

Let's say we have `void x() {}` declared, so `x()` has static type void.

Can I do `x().foo();`? That's prevented by the current semantics, so probably stays that way.

Can I do `V(x()).foo();`? That's probably allowed (the context type for `x()` is `void`, and we allow assigning `void` to `void`. Then we're inside `V`, and calling `foo` is fine. Inside `V`, the static type of `this` will be void, which puts a dampener on the possibilities.

Can I do `V y = x(); y.foo();`? Maybe not. We're assigning `void` to something which is not `void` or `dynamic`.

Can I do `V y = V(x()); y.foo();`? Probably, the `V(x())` is assigning `void` to `void`, and then we're out of the `void` type.

What if `V` was declared as `extension V<T> on T { int foo() => 42; }`.
The `x().foo()` and `V<void> y = x();` would still both be disallowed, and `V(x())` would still be allowed after inference to `V<void>(x())`. Inside the `V`, the type of `T` would not be `void`, but that's a known issue with generics, nothing new here.
So, all in all, seems like things work like they should.

What about `dynamic`. I guess we get dynamic downcast to non-protected extension types, and no downcast to protected extension types because we disallow explicit casts too.

Reply:
Nice analysis! I think I agree with all the points made here, but I also think the whole idea about defining `void` as an extension type would need some further scrutiny.

Reply:
I've rewritten the part about `void` to indicate the known problems.